File: src/main/java/openmedstack/commands/DomainCommand.java
package openmedstack.commands;

import openmedstack.ICorrelate;

import java.time.Instant;

public abstract class DomainCommand implements ICorrelate {
    private final Instant _timestamp;
    private final String _aggregateId;
    private final Integer _version;
    private final String _correlationId;

    protected DomainCommand(String aggregateId, Integer version, Instant timeStamp, String correlationId) {
        if (timeStamp == Instant.MIN) {
            throw new IllegalArgumentException("Cannot use min time");
        }
        _aggregateId = aggregateId;
        _version = version;
        _timestamp = timeStamp;
        _correlationId = correlationId;
    }

    public String getAggregateId() {
        return _aggregateId;
    }

    public Integer getVersion() {
        return _version;
    }

    public Instant getTimestamp() {
        return _timestamp;
    }

    public String getCorrelationId() {
        return _correlationId;
    }
}

File: modules/openmedstack/src/main/kotlin/org/openmedstack/commands/DomainCommand.kt
package org.openmedstack.commands

import org.openmedstack.ICorrelate
import java.time.OffsetDateTime

abstract class DomainCommand protected constructor(
    aggregateId: String,
    version: Int,
    timeStamp: OffsetDateTime,
    correlationId: String?
) : ICorrelate {
    private val _timestamp: OffsetDateTime
    private val _aggregateId: String
    private val _version: Int
    private val _correlationId: String?
    override val correlationId: String?
        get() = _correlationId
    val timestamp: OffsetDateTime
        get() = _timestamp
    val aggregateId: String
        get() = _aggregateId
    val version: Int
        get() = _version

    init {
        require(timeStamp != OffsetDateTime.MIN) { "Cannot use min time" }
        _aggregateId = aggregateId
        _version = version
        _timestamp = timeStamp
        _correlationId = correlationId
    }
}
